app/models/concerns/json_schema_validator.rb

```diff
@@ -0,0 +1,97 @@
+# This file defines a custom validator class `JsonSchemaValidator` for validating a JSON object against a schema.
+# To use this validator, define a schema as a Ruby hash and include it in the validation options when validating a model.
+# The schema should define the expected structure and types of the JSON object, as well as any validation rules.
+# Here's an example schema:
+#
+# schema = {
+#   'name' => { required: true, type: 'string' },
+#   'age' => { required: true, type: 'integer' },
+#   'is_active' => { required: false, type: 'boolean' },
+#   'tags' => { required: false, type: 'array' },
+#   'address' => {
+#     required: false,
+#     type: 'hash',
+#     properties: {
+#       'street' => { required: true, type: 'string' },
+#       'city' => { required: true, type: 'string' }
+#     }
+#   }
+# }.freeze
+#
+# To validate a model using this schema, include the `JsonSchemaValidator` in the model's validations and pass the schema
+# as an option:
+#
+# class MyModel < ApplicationRecord
+#   validates_with JsonSchemaValidator, schema: schema
+# end
+
+class JsonSchemaValidator < ActiveModel::Validator
+  def validate(record)
+    # get the attribute resolver function from options or use a default one
+    attribute_resolver = options[:attribute_resolver] || ->(rec) { rec.additional_attributes }
+
+    # resolve the JSON data to be validated
+    json_data = attribute_resolver.call(record)
+
+    # get the schema to be used for validation
+    schema = options[:schema]
+
+    # call the private method to validate the schema
+    validate_schema(json_data, schema, 'root', record)
+  end
+
+  private
+
+  def validate_schema(json_data, schema, path, record)
+    schema.each do |key, rule|
+      value = json_data[key] # get the value for the current key
+
+      # check if the value is required and not present
+      if rule[:required] && value.nil?
+        record.errors.add(path, "#{key} is required")
+      elsif !value.nil? # check if the value is present
+        validate_type(value, rule[:type], "#{path}.#{key}", record) # validate the type of the value
+
+        # validate the nested schema if the value is a Hash
+        validate_schema(value, rule[:properties], "#{path}.#{key}", record) if rule[:properties] && value.is_a?(Hash)
```

Comment:
@scmmishra seems like we already have the `JSONSchemer` gem included in the code base to validate with JSONSchema, so some of the logic here would be redundant. 

Example: https://github.com/chatwoot/chatwoot/blob/5cdc7d654a461d9b20d583e19ed01e33a31f202d/app/models/integrations/hook.rb#L57